lib/core/context.c

```diff
@@ -190,3 +190,69 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
     h2o_time2str_rfc1123(ctx->_timestamp_cache.value->rfc1123, &gmt);
     h2o_time2str_log(ctx->_timestamp_cache.value->log, ctx->_timestamp_cache.tv_at.tv_sec);
 }
+
+void h2o_context_close_idle_connections(h2o_context_t *ctx, size_t max_connections_to_close, uint64_t min_age)
+{
+    if (max_connections_to_close <= 0)
+        return;
+
+    size_t closed = ctx->_conns.num_conns.shutdown;
+
+    if (closed >= max_connections_to_close)
+        return;
+
+    H2O_CONN_LIST_FOREACH(h2o_conn_t * conn, ({&ctx->_conns.idle}), {
+        struct timeval now = h2o_gettimeofday(ctx->loop);
+        if (h2o_timeval_subtract(&conn->connected_at, &now) < (min_age * 1000))
+            continue;
+        ctx->connection_stats.idle_closed++;
+        conn->callbacks->close_idle_connection(conn);
+        closed++;
+        if (closed == max_connections_to_close)
+            return;
+    });
+}
+
+static size_t *get_connection_state_counter(h2o_context_t *ctx, h2o_conn_state_t state)
+{
+    return ctx->_conns.num_conns.counters + (size_t)state;
+}
+
+static void _connection_state_set(h2o_conn_t *conn, h2o_conn_state_t state)
+{
+    if (h2o_linklist_is_linked(&conn->_conns)) {
+        h2o_linklist_unlink(&conn->_conns);
+    }
+    switch (state) {
+    case H2O_CONN_STATE_IDLE:
+        h2o_linklist_insert(&conn->ctx->_conns.idle, &conn->_conns);
+        break;
+    case H2O_CONN_STATE_ACTIVE:
+        h2o_linklist_insert(&conn->ctx->_conns.active, &conn->_conns);
+        break;
+    case H2O_CONN_STATE_SHUTDOWN:
+        h2o_linklist_insert(&conn->ctx->_conns.shutdown, &conn->_conns);
+        break;
+    }
+}
+
+void h2o_conn_init_state(h2o_conn_t *conn, h2o_conn_state_t state)
+{
+    conn->state = state;
+    ++*get_connection_state_counter(conn->ctx, conn->state);
+    _connection_state_set(conn, state);
+}
+
+void h2o_conn_fin_state(h2o_conn_t *conn)
+{
+    --*get_connection_state_counter(conn->ctx, conn->state);
+    h2o_linklist_unlink(&conn->_conns);
+}
+
+void h2o_conn_set_state(h2o_conn_t *conn, h2o_conn_state_t state)
+{
+    --*get_connection_state_counter(conn->ctx, conn->state);
+    conn->state = state;
+    ++*get_connection_state_counter(conn->ctx, conn->state);
+    _connection_state_set(conn, state);
```

Comment:
This is nothing more than a nitpick, but it might make sense to create two internal functions:
* one that increments the counter and links the connection, and
* another that decrements the counter and unlinks the connection.

Then, we can call the former in `h2o_conn_init_state`, call the latter in `h2o_conn_fin_state`, and call both in `h2o_conn_set_state`.